third_party/move/move-vm/runtime/src/storage/loader.rs

```diff
@@ -0,0 +1,291 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    config::VMConfig,
+    loader::{Function, LoadedFunctionOwner, Module, TypeCache},
+    logging::expect_no_verification_errors,
+    module_linker_error,
+    module_traversal::TraversalContext,
+    storage::{
+        environment::RuntimeEnvironment, module_storage::ModuleStorage,
+        struct_name_index_map::StructNameIndexMap,
+    },
+    CodeStorage, LoadedFunction,
+};
+use move_binary_format::{
+    access::{ModuleAccess, ScriptAccess},
+    errors::{Location, PartialVMError, PartialVMResult, VMResult},
+    CompiledModule,
+};
+use move_core_types::{
+    account_address::AccountAddress,
+    gas_algebra::NumBytes,
+    identifier::IdentStr,
+    language_storage::{ModuleId, TypeTag},
+    vm_status::StatusCode,
+};
+use move_vm_types::{
+    gas::GasMeter,
+    loaded_data::runtime_types::{StructType, Type, TypeBuilder},
+};
+use parking_lot::RwLock;
+use std::{collections::BTreeMap, sync::Arc};
+use typed_arena::Arena;
+
+/// V2 implementation of loader, which is stateless - i.e., it does not contain
+/// module or script cache. Instead, module and script storages are passed to all
+/// APIs by reference.
+pub(crate) struct LoaderV2 {
+    runtime_environment: Arc<RuntimeEnvironment>,
+    // Local caches:
+    //   These caches are owned by this loader and are not affected by module
+    //   upgrades. When a new cache is added, the safety guarantees (i.e., why
+    //   it is safe for the loader to own this cache) MUST be documented.
+    // TODO(loader_v2): Revisit type cache implementation. For now re-use the existing
+    //                  one to unblock upgradable module and script storage first.
+    ty_cache: RwLock<TypeCache>,
+}
+
+impl LoaderV2 {
+    pub(crate) fn new(runtime_environment: Arc<RuntimeEnvironment>) -> Self {
+        Self {
+            runtime_environment,
+            ty_cache: RwLock::new(TypeCache::empty()),
+        }
+    }
+
+    pub(crate) fn runtime_environment(&self) -> &RuntimeEnvironment {
+        self.runtime_environment.as_ref()
+    }
+
+    pub(crate) fn vm_config(&self) -> &VMConfig {
+        self.runtime_environment.vm_config()
+    }
+
+    pub(crate) fn ty_builder(&self) -> &TypeBuilder {
+        &self.vm_config().ty_builder
+    }
+
+    pub(crate) fn ty_cache(&self) -> &RwLock<TypeCache> {
+        &self.ty_cache
+    }
+
+    pub(crate) fn struct_name_index_map(&self) -> &StructNameIndexMap {
```

Comment:
how does this work if we need to invalidate? this is still an append only data structure right? 

are we only making published modules visible during rolling commit? unlike resource writes becomes visible immediately after execution

Reply:
Yes, published modules are only visible after rolling commit. For init module, we start a new VM so that we don't cache what needs invalidation later.

Btw, I realised new VM makes V2 incompatible with V1 because finishing session before init module messes up gas charges for cached resources. I will move struct name index map (runtime env) to block executor and VM should not know about it. This should be not many changes.

Reply:
wondering how do we support module eviction with this struct later if we have a long running base cache for modules

Reply:
For every evicted module, once can go over it struct types and evict the entries in name cache, but that would need to keep some reference counting I guess because there can be other modules using it.